apps/src/lib/levelbuilder/MarkdownEnabledTextarea.jsx

```diff
@@ -0,0 +1,131 @@
+    margin: 0
+  },
+  icon: {
+    marginRight: 7
+  }
+};
+
+export const markdownFeaturesShape = PropTypes.shape({
+  imageUpload: PropTypes.bool,
+  resourceLink: PropTypes.bool
+});
+
+export default class MarkdownEnabledTextarea extends React.Component {
+  static propTypes = {
+    markdown: PropTypes.string,
+    name: PropTypes.string,
+    inputRows: PropTypes.number,
+    handleMarkdownChange: PropTypes.func.isRequired,
+    features: markdownFeaturesShape
+  };
+
+  static defaultProps = {
+    features: {}
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {};
+  }
+
+  changeMarkdown(newMarkdown) {
+    let e = {target: {value: newMarkdown}};
+    this.props.handleMarkdownChange(e);
+  }
+
+  handleOpenUploadImage = () => {
+    this.setState({uploadImageOpen: true});
+  };
+
+  handleOpenAddResourceLink = () => {
+    this.setState({addResourceLinkOpen: true});
+  };
+
+  handleCloseUploadImage = () => {
+    this.setState({uploadImageOpen: false});
+  };
+
+  handleCloseAddResourceLink = () => {
+    this.setState({addResourceLinkOpen: false});
+  };
+
+  handleConfirmAddResourceLink = resourceKey => {
+    if (resourceKey) {
+      this.changeMarkdown(this.props.markdown + `\n\n[r ${resourceKey}]`);
+    }
+    this.handleCloseAddResourceLink();
+  };
+
+  handleUploadImage = url => {
+    this.changeMarkdown(this.props.markdown + `\n\n![](${url})`);
+  };
+
+  render() {
+    return (
+      <div>
+        <div style={{margin: 5}}>
+          <textarea
+            name={this.props.name}
+            onChange={this.props.handleMarkdownChange}
+            rows={this.props.inputRows || 5}
+            style={styles.input}
+            value={this.props.markdown}
+          />
+          {this.props.features.imageUpload && (
+            <button
+              className="btn"
+              onClick={this.handleOpenUploadImage}
+              type="button"
+            >
+              <i style={styles.icon} className="fa fa-plus-circle" />
+              Image
+            </button>
+          )}
+          {this.props.features.resourceLink && (
+            <button
+              className="btn"
+              onClick={this.handleOpenAddResourceLink}
+              type="button"
+            >
+              <i style={styles.icon} className="fa fa-plus-circle" />
+              Resource
+            </button>
+          )}
+        </div>
+        {this.props.features.imageUpload && (
+          <UploadImageDialog
+            handleClose={this.handleCloseUploadImage}
+            isOpen={!!this.state.uploadImageOpen}
+            uploadImage={this.handleUploadImage}
+          />
+        )}
+        {this.props.features.resourceLink && (
+          <FindResourceDialog
+            handleClose={this.handleCloseAddResourceLink}
+            handleConfirm={this.handleConfirmAddResourceLink}
+            isOpen={!!this.state.addResourceLinkOpen}
+          />
+        )}
+      </div>
+    );
+  }
+}
```

Comment:
Sorry for the late request, but would you be willing to do some git-fu like `git mv TextareaWithImageUpload.jsx MarkdownEnabledTextarea.jsx` to preserve revision history and make diffs easier to see, now and in the future? (here and below for test file) Open to reasons we might not want to do this, but would prefer this pattern if there is no reason not to.

Reply:
That's already the state of the current revision history. There were just enough edits made to the files after moving them that they show up as different files in the overall diff, but if you examine the diff for the first commit individually you'll see what you're looking for